serde-reflection/src/trace.rs

```diff
@@ -245,6 +285,26 @@ impl Tracer {
         }
     }
 
+    /// Same as `trace_type_once` but if any uncovered variants remain in the
+    /// recursive format, we repeat the process.
+    /// We accumulate and return all the sampled values at the end.
+    pub fn trace_type_all_variants<'de, T>(
+        &mut self,
+        samples: &'de Samples,
+    ) -> Result<(Format, Vec<T>)>
+    where
+        T: Deserialize<'de>,
+    {
+        let mut values = Vec::new();
+        loop {
+            let (format, value) = self.trace_type_once::<T>(samples)?;
+            values.push(value);
+            if self.incomplete_enums.is_empty() {
```

Comment:
Has this strategy been working in your project? I've also worked to solve the incomplete enum issue as it's a huge pain point in our project but IIRC (it's been a few months) an arrangement like `enum -> struct -> enum` can result in one variant of the top enum being traced which leads it to tracing the struct, then the outermost enum which results in an infinite loop on the outermost enum due to the top enum being in `incomplete_enums`.

I'm excited about any progress on this front, I'll try to pull your branch later and see if it can handle some of our complex data types.

The approach I had been working on was to trace from top to bottom based on index, one variant at a time and only move forward on a parent type once its child was completed. If the type == Self then consider it complete. In other words:
``` rust
enum Enum1 {
  One(Foo),
  Two
}

struct Foo {
  baz: Enum2,
  bat: i32,
}

enum Enum2 {
  One,
  Two
}
```
would be traced like:
```
Enum1 -> One(Foo) -> Foo -> baz -> Enum2 -> One
Enum1 -> One(Foo) -> Foo -> baz -> Enum2 -> Two
Enum1 -> One(Foo) -> Foo -> bat
Enum1 -> Two
```

Reply:
Good point!

This approach was sufficient in my use case but (I haven’t tested this yet but my gut feeling after checking the code) is that such a case with an outer enum would loop indefinitely. Detecting when a type has been fully traced (once all of its fields / variants have been fully traced) would be a good approach, though quite a complex one. Please share your approach (I’m happy to collaborate on this PR) so we can advance our combined use case :)

I think that in any case this PR is an initial step forward. If tracing any type successfully is right around the corner it could be skipped. However, that seems to be a really hard problem where there might for quite a while be some cases where tracing may not terminate. Therefore, I think that offering this opt-in API is a good compromise that allows users to disable the will-definitely-terminate-but-may-not-trace-everything guarantee if they know that their use case can be traced exhaustively with termination using the existing rules.

Reply:
TLDR I hope that the tracing algorithm gradually improves to reach more cases while guaranteeing termination. Until it can trace every type and terminate, this minor API addition offers an opt-in if the user knows that tracing will definitely terminate in their use case.